lib/Controller/RoomController.php

```diff
@@ -1222,23 +1222,28 @@ public function addParticipantToRoom(string $newParticipant, string $source = 'u
 			return new DataResponse([], Http::STATUS_BAD_REQUEST);
 		}
 
-		$participants = $this->participantService->getParticipantUserIds($this->room);
+		$participants = $this->participantService->getParticipantsForRoom($this->room);
+		$participantsByUserId = [];
+		foreach ($participants as $participant) {
+			if ($participant->getAttendee()->getActorType() === Attendee::ACTOR_USERS) {
+				$participantsByUserId[$participant->getAttendee()->getActorId()] = $participant;
+			}
+		}
 
+		// list of participants to attempt adding,
+		// existing ones will be filtered later below
 		$participantsToAdd = [];
+
 		if ($source === 'users') {
 			$newUser = $this->userManager->get($newParticipant);
 			if (!$newUser instanceof IUser) {
 				return new DataResponse([], Http::STATUS_NOT_FOUND);
 			}
 
-			if (\in_array($newParticipant, $participants, true)) {
-				return new DataResponse([]);
-			}
-
-			$this->participantService->addUsers($this->room, [[
+			$participantsToAdd[] = [
```

Comment:
am thinking to make this a simple user id array and adding the actorType only at the end

Reply:
But not all participantsToAdd are users? There are already emails for guests, and we will have groups and circles tracking at some point too.

Reply:
> But not all participantsToAdd are users?
In the current implementation it is, I simply refactored the code to have the adding logic at the end in a common location to avoid duplicating the existence check. Please check further below.

One exception indeed is the email logic, but that one is not adding any user and has its own block